### DISCUSSION TITLE
[Proposal]: Introduce an operator like TypeScript's Omit utility type.

### DISCUSSION THREAD

**Discussion by liamwang** [dotnet/csharplang#5862] [Language Ideas]
# Omit operator

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary

Introduce an operator that allows a class to "inherit" from another class while excluding specified properties. Just like TypeScript's `Omit` utility type.

## Motivation

Many times I want to define a few model classes that inherit from another class, but exclude some properties. For example, there is a class like this:

```cs
public class User : IEntity
{
    public long Id { get; set; }

    public string Name { get; set; }

    // Other properties here...

    public string PasswordHash { get; set; }

    public ICollection<UserRole> UserRoles { get; set; }
}
```

Then I want to define a few other models, like this:

```cs
public class UserDto
{
    public long Id { get; set; }

    public string Name { get; set; }

    //  Here are the other properties, but not `PasswordHash ` and `UserRoles`
}
```

If one day, I add, delete or modify the properties in `User` class, it is likely that other classes will have to be modified accordingly.

Expecting an **`omit`** operator, which might be used like this:

```diff
public class UserDto 
+    : User omit [PasswordHash, UserRoles]
{
}
```

This way, when the properties of the "parent" class change, you don't have to manually modify other classes or worry about missing something.

**Comment by CyrusNajmabadi** [comment]
Converted this to a discussion as per our rules about creating language proposals as issues.

**Comment by TahirAhmadov** [comment]
I can see the value in this, but what happens in the IL/metadata? I doubt it's possible to make this work, or to even change the CLR to make this work.

**Comment by HaloFour** [comment]
Feels like this could be a use case for a source generator, one that could copy the schema of another type but offer the ability to specify which members should not be included.

The developer would write:

```cs
[BasedOn<User>(Omits = { "PasswordHash", "UserRoles" })]
public partial class UserDto { }
```

and the generator would emit:

```cs
public partial class UserDto {
    public long Id { get; set; }
    public string Name { get; set; }
}
```

The source generator could also emit members like a copy constructor or conversion operators from `User`, etc.

**Reply by TahirAhmadov** [reply]
I was thinking the same thing, to make it work with "native" inheritance is pretty much impossible, and if the objective is to just copy a few properties, your idea should work.

**Comment by HZ-GeLiang** [comment]
I implemented a Demo based on Source Generator.It only works in VS2022 and NET Core.
https://gitee.com/hzgel/POCOSourceGenerator
